### DISCUSSION TITLE
Resource usage

### DISCUSSION THREAD

**Discussion by nfettinger** [thirtythreeforty/neolink#158] [Q&A] [answered]
If I run this on an Ubuntu Server platform as a service, what sort of CPU and memory resources get consumed? Can a RPI4 support running this? What sort of frame delay does this add to the stream?

I am hoping to run this on the same 18.04 server as the NVR software, so hopefully a loopback adapter can be used to bridge the two. If this is not possible, is it safe to assume the network bandwidth is basically equal to the stream from the camera?

**Comment by QuantumEntangledAndy** [comment, marked as answer]
so I run this on my own rpi3 so a 4 should be fine. The resources on the ram and CPU are not much but I don't have exact numbers on hand. 

All we are doing is pulling frames from the camera filtering out the packets the camera gives us into audio and video and then passing the video to gstreamer for it to serve as rtsp.

We do however do some processing on the audio because some of the cameras use an old audio format that is not supported for use over rtsp.

The major requirment is bandwidth. Don't forget that you are going to be both receiving and sending a HD stream at once so it will probably be double what you expect.

There also seems to be a limit to how many open connections neolink can handle at once which tends to be about 5-6 HD streams at once. I think at this point there is so much traffic coming into the port that it has trouble receiving it all at once.

Hope that is helpful.

**Reply by nfettinger** [reply]
Just as an FYI to anyone who has this question in the future (or has issues with bandwidth):

According to [raspberrypi.org](https://magpi.raspberrypi.org/articles/raspberry-pi-4-specs-benchmarks), even though some models have a gigabit Ethernet port, the bus bandwidth may be shared with other SOC devices (including USB). This yielded a max of 237 Mbps on a RPI3B+ compared to the 943 Mbps on a RPI4B.

Further Reading (Full vs Half Duplex):
The Pi Ethernet ports advertise full duplex, meaning traffic can flow at the advertised rate of 1Gbps simultaneously on both the uplink (transmit) and downlink (receive). Basically, you should be able to accept data at 1Gbps and send it out at the same rate.
However, other 3rd party benchmarks has shown that the bandwidth can be asymmetrical due to restricted resources; one showing the RPI4 can accept data at the max rate or 1Gbps, but can only output 100Mbps when the CPU was taxed with other processing intensive tasks.

Therefore, your mileage may vary.
